flowauth/backend/flowauth/roles.py

```diff
@@ -0,0 +1,162 @@
+            "%Y-%m-%dT%H:%M:%S.%fZ"
+        ),
+        "longest_token_life_minutes": role.longest_token_life_minutes,
+        "server": role.server_id,
+        "users": sorted([user.id for user in role.users]),
+    }
+
+
+def roles_to_json(roles):
+    return jsonify(
+        sorted([role_to_dict(role) for role in roles], key=lambda x: x["id"])
+    )
+
+
+@blueprint.route("/", methods=["GET"])
+@login_required
+@admin_permission.require(http_exception=401)
+def list_roles():
+    return jsonify([role_to_dict(role) for role in Role.query.all()])
+
+
+@blueprint.route("/user/<user_id>")
+@login_required
+@admin_permission.require(http_exception=401)
+def list_user_roles(user_id):
+    user = User.query.filter(User.id == user_id).first_or_404()
+    return jsonify([role_to_dict(role) for role in user.roles])
+
+
+@blueprint.route("/<role_id>", methods=["GET"])
+@login_required
+@admin_permission.require(http_exception=401)
+def get_role(role_id):
+    role = Role.query.filter(Role.id == role_id).first()  # First or error?
+    return jsonify(role_to_dict(role))
+
+
+@blueprint.route("/", methods=["POST"])
+@login_required
+@admin_permission.require(http_exception=401)
+def add_role():
+    json = request.get_json()
+    json["latest_token_expiry"] = datetime.datetime.strptime(
+        json["latest_token_expiry"], "%Y-%m-%dT%H:%M:%S.%fZ"
+    )
+    server = Server.query.filter(Server.id == json["server_id"]).first()
+    role_scopes = Scope.query.filter(Scope.id.in_(json["scopes"])).all()
+    try:
+        role_users = User.query.filter(User.id.in_(json["users"])).all()
+    except KeyError:
+        role_users = []
+    new_role = Role(
+        name=json["name"],
+        scopes=role_scopes,
+        users=role_users,
+        server=server,
+        latest_token_expiry=json["latest_token_expiry"],
+        longest_token_life_minutes=json["longest_token_life_minutes"],
+    )
+    db.session.add(new_role)
+    db.session.commit()
+    current_app.logger.info(f"Created role {new_role.name}")
+    return get_role(new_role.id)
+
+
+# NOTES FOR REVIEW: I wondered whether to have this merge lists of users and scopes
+# when updated, but in the end I think that it's cleaner to have that on the frontend
+# and make an explicit decision
+@blueprint.route("/<role_id>", methods=["PATCH"])
+@login_required
+@admin_permission.require(http_exception=401)
+def edit_role(role_id):
+    edits = request.get_json()
+    role = Role.query.filter(Role.id == role_id).first_or_404()
+    for key, value in edits.items():
+        if key == "id":
+            current_app.logger.warning("Cannot change role ID; ignoring")
+            pass
+        if key == "users":
+            value = [User.query.filter(User.id == uid).first() for uid in value]
+        elif key == "scopes":
+            value = [Scope.query.filter(Scope.id == sid).first() for sid in value]
+        elif key == "server":
+            value = Server.query.filter(Server.id == value).first()
```

Comment:
I don't think this should allow editing the server for a role.  A role should be associated with a server on creation, and I don't think it would ever make sense to change this (not least because a different server may have a different set of scopes, so if you change the server for a previously-defined role then there would need to be a messy process of checking which of the role's scopes are still applicable for the new server).